multiplex.go

```diff
@@ -567,6 +572,30 @@ func (mp *Multiplex) readNext() ([]byte, error) {
 	return buf[:n], nil
 }
 
+func (mp *Multiplex) getBufferInbound(length int) ([]byte, error) {
+	select {
+	case mp.bufIn <- struct{}{}:
+	case <-mp.shutdown:
+		return nil, ErrShutdown
+	}
+
+	return mp.getBuffer(length)
+}
+
+func (mp *Multiplex) getBufferOutbound(length int, timeout, cancel <-chan struct{}) ([]byte, error) {
+	select {
+	case mp.bufOut <- struct{}{}:
+	case <-timeout:
+		return nil, errTimeout
+	case <-cancel:
+		return nil, ErrStreamClosed
+	case <-mp.shutdown:
+		return nil, ErrShutdown
+	}
+
+	return mp.getBuffer(length)
+}
+
 func (mp *Multiplex) getBuffer(length int) ([]byte, error) {
 	if err := mp.memoryManager.ReserveMemory(length, 128); err != nil {
```

Comment:
Can we reserve all the memory mplex could possibly consume up front, and get rid of this call?

Reply:
we could, but thats 16M a pop!!! 
16 buffers x 1M max message size.

But you are right, we should.
I guess we can rsvp incrementally to find how much we can wiggle, up to 16 bufs (maybe less?)

Reply:
ok done; I also cut it to 8 buffers total (4 each direction) so that it's only up to 8M for the reservation.